interceptor/middleware/logging.go

```diff
@@ -0,0 +1,85 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+
+	"github.com/kedacore/http-add-on/pkg/util"
+)
+
+const (
+	CombinedLogFormat     = `%s %s %s [%s] "%s %s %s" %d %d "%s" "%s"`
+	CombinedLogTimeFormat = "02/Jan/2006:15:04:05 -0700"
+	CombinedLogBlankValue = "-"
+)
+
+type Logging struct {
+	logger          logr.Logger
+	upstreamHandler http.Handler
+}
+
+func NewLogging(logger logr.Logger, upstreamHandler http.Handler) *Logging {
+	return &Logging{
+		logger:          logger,
+		upstreamHandler: upstreamHandler,
+	}
+}
+
+var _ http.Handler = (*Logging)(nil)
+
+func (lm *Logging) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	r = util.RequestWithLogger(r, lm.logger.WithName("LoggingMiddleware"))
+	w = newLoggingResponseWriter(w)
+
+	var sw util.Stopwatch
+	defer lm.logAsync(w, r, &sw)()
+
+	sw.Start()
+	defer sw.Stop()
+
+	lm.upstreamHandler.ServeHTTP(w, r)
+}
+
+func (lm *Logging) logAsync(w http.ResponseWriter, r *http.Request, sw *util.Stopwatch) func() {
+	signaler := util.NewSignaler()
+
+	go lm.log(w, r, sw, signaler)
+
+	return func() {
+		go signaler.Signal()
+	}
```

Comment:
Could we have perf problems creating 2 routines for each request? Can't we just avoid the signaler doing something like:
```golang
return func() {
		go lm.log(w, r, sw)
	}
```

Or even if maybe making it sync

Reply:
I made this async to avoid adding extra latency to the request. Logs are generally slow to write, so I don't like the idea of doing this synchronously.

I'm not sure how long does it take to invoke `go`, so I'll take a better look at this. Regarding the runtime impact of the new Go Routine, it will stay parked for the majority of its life, so that's negligible.

Reply:
The alternative where we just use a single Go Routine would look more like:
```go
	t0 := time.Now()

	return func() {
		d := time.Since(t0)
		go lm.log(w, r, sw, d)
	}
```

That actually seems a better implementation, as we would have more accurate timing and I don't think the calls to `time.Now` and `time.Since` would impact the latency significantly.

Reply:
so, will you update the implementation?

Reply:
Yes, I will be doing that

Reply:
I reread the implementation and noticed we already have the Stopwatch instance there. So your original code suggestion was perfect, there's no need to call time.Now/Since.

Reply:
This is now patched